File: tools/validate_json.py
"""Validate every JSON document shipped with the game.

The old validator only inspected files directly below ``data/``. Most game
content lives in nested catalog folders, so malformed production data could
pass the default quality gate. Keep this script dependency-free because it is
used locally and in GitHub Actions.
"""

from __future__ import annotations

import json
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
JSON_ROOTS = (ROOT / "data", ROOT / "schemas")


def json_files() -> list[Path]:
    return sorted(
        path
        for folder in JSON_ROOTS
        if folder.exists()
        for path in folder.rglob("*.json")
        if path.is_file()
    )


def main() -> int:
    paths = json_files()
    if not paths:
        print("FAIL no JSON files found in data/ or schemas/")
        return 1

    failures: list[str] = []
    for path in paths:
        relative = path.relative_to(ROOT)
        try:
            json.loads(path.read_text(encoding="utf-8"))
            print("OK", relative)
        except (OSError, UnicodeError, json.JSONDecodeError) as exc:
            failures.append(f"{relative}: {exc}")
            print("FAIL", relative, exc)

    print(f"Validated {len(paths)} JSON file(s); failures={len(failures)}")
    return 1 if failures else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/node/validate_json.mjs
import { readFile, readdir } from "node:fs/promises";
import { join, relative } from "node:path";

const folders = ["data", "schemas"];
let checked = 0;
let failed = false;

const walkJson = async (folder) => {
  const entries = await readdir(folder, { withFileTypes: true }).catch(() => []);
  const nested = await Promise.all(
    entries.map(async (entry) => {
      const path = join(folder, entry.name);
      if (entry.isDirectory()) return walkJson(path);
      return entry.isFile() && entry.name.endsWith(".json") ? [path] : [];
    }),
  );
  return nested.flat();
};

for (const folder of folders) {
  const files = (await walkJson(folder)).sort();
  for (const filePath of files) {
    checked += 1;
    try {
      JSON.parse(await readFile(filePath, "utf8"));
      console.log(`OK ${relative(".", filePath)}`);
    } catch (error) {
      failed = true;
      console.error(`INVALID ${relative(".", filePath)}: ${error.message}`);
    }
  }
}

if (checked === 0) {
  console.error("No JSON files were found in data/ or schemas/.");
  process.exit(1);
}

if (failed) process.exit(1);
console.log(`Validated ${checked} JSON file(s).`);
